sound/soc/sof/intel/cnl.c

```diff
@@ -161,11 +161,9 @@ static void cnl_ipc_dsp_done(struct snd_sof_dev *sdev)
 static bool cnl_compact_ipc_compress(struct snd_sof_ipc_msg *msg,
 				     u32 *dr, u32 *dd)
 {
-	struct sof_ipc_pm_gate *pm_gate;
-
-	if (msg->header == (SOF_IPC_GLB_PM_MSG | SOF_IPC_PM_GATE)) {
-		pm_gate = msg->msg_data;
+	struct sof_ipc_pm_gate *pm_gate = msg->msg_data;
 
+	if (pm_gate->hdr.cmd == (SOF_IPC_GLB_PM_MSG | SOF_IPC_PM_GATE)) {
```

Comment:
originally we were reading the generic `msg` pointer here, which is guaranteed to point to a valid `struct snd_sof_ipc_msg` instance, and only if we find out, that this is a `PM_GATE` message, then we treat it as such, i.e. we treat its data as `struct sof_ipc_pm_gate *pm_gate`. Now you immediately assume, that `msg->msg_data` points to `struct snd_sof_ipc_msg`, but that might not be the case? Otherwise we wouldn't need to check `cmd` at all

Reply:
@lyakh, the `msg->msg_data` must be always points to an ipc message and an ipc message must contain at leas the `struct snd_sof_ipc_msg`.
If we would have a single place where we are not providing the message for tx then I would have added couple of more patches to do this change in a more gradual way, but we _always_ give a message to tx and the msg->header is _always_ set to hdr->cmd.

I have heard that in the distant past SOF supported messages without mailbox (via purely registers), but we no longer have that (IPC4 is a different matter, this is not IPC4).

Reply:
it's not just a distant past, for specific IPC functions we DO use registers only because the memory windows may not be powered. That's the reason why power gating commands are NOT using the mailbox.

So now I wonder if this change is compatible with the pm_gate IPCs used for D0ix support?

Reply:
I think the change is OK and will nto break D0ix support but it isnt as intuitive as before. I think the check for msg->header to ensure it is a pm gate message made more sense

Reply:
> 

@plbossart, it will work (and works) as it used to. We always given the IPC message and never had a case where the msg was NULL and the header parameter used only.